language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
+                // the current set includes everything in `to_add` we have seen so far
+                assert forall j in 0..i: contains(members, to_add[j]);
+                // having no new members means that all elements in the `to_add` vector we have seen so far are already
+                // in the existing set, and vice versa.
+                assert len(members) == len(old(members)) <==> (forall j in 0..i: contains(old(members), to_add[j]));
+            };
+            (i < num_to_add)
+        }) {
+            let entry = Vector::borrow(to_add, i);
+            if (!Vector::contains(members, entry)) {
+                Vector::push_back(members, *entry);
+            };
+            i = i + 1;
         };
 
-        exist
+        Vector::length(members) > num_existing
+    }
+    spec add_members_internal {
+        pragma opaque;
+        // TODO(mengxu): this is to force the prover to honor the "opaque" pragma in the ignore opaque setting
+        aborts_if [abstract] false;
+        aborts_if [concrete] false;
+        include AddMembersInternalEnsures<T> {
+            old_members: old(members),
+            new_members: members,
+        };
+        // ensures that the `members` argument is and remains a set
+        invariant forall i in 0..len(members), j in 0..len(members):
+            members[i] == members[j] ==> i == j;
+        // returns whether a new element is added to the set
+        ensures result == (exists i in 0..len(to_add): !contains(old(members), to_add[i]));
     }
+    spec schema AddMembersInternalEnsures<T: copy> {
+        old_members: vector<T>;
+        new_members: vector<T>;
+        to_add: vector<T>;
+        // everything in the `to_add` vector must be in the updated set
+        ensures forall i in 0..len(to_add): contains(new_members, to_add[i]);
+        // everything in the old set must remain in the updated set
+        ensures forall i in 0..len(old_members): contains(new_members, old_members[i]);
+        // everything in the updated set must come from either the old set or the `to_add` vector
+        ensures forall i in 0..len(new_members): (contains(old_members, new_members[i]) || contains(to_add, new_members[i]));
+    }
+
+    /// Remove all elements that appear in `to_remove` from `members`.
+    ///
+    /// The `members` argument is essentially a set simulated by a vector, hence
+    /// the uniqueness of its elements are guaranteed, before and after the bulk
+    /// removal. The `to_remove` argument, on the other hand, does not guarantee
+    /// to be a set and hence can have duplicated elements.
+    fun remove_members_internal<T: drop>(
+        members: &mut vector<T>,
+        to_remove: &vector<T>,
+    ): bool {
+        let num_existing = Vector::length(members);
+        let num_to_remove = Vector::length(to_remove);
 
-    spec remove_identity {
+        let i = 0;
+        while ({
+            spec {
+                assert i <= num_to_remove;
```

Comment:
I'm assuming it's `<=` because it comes before the condition

Reply:
Yes.

A **loop invariant** is a property of a program loop that is true at every iteration **before** hitting the loop body. And the loop breaking condition (`i < num_to_remove`) is considered as part of the loop body.